File: .leetcode/33.搜索旋转排序数组.js
/*
 * @lc app=leetcode.cn id=33 lang=javascript
 *
 * [33] 搜索旋转排序数组
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var search = function(nums, target) {
    // 先转为有序数组
    const revPos = findRevPos(nums, 0, nums.length);
    const right = nums.splice(0,revPos+1);
    nums = nums.concat(right);
    // 二分查找
    const newPos = findPos( nums, target, 0, nums.length );
    const orgRightLength = nums.length - revPos - 1;
    if( revPos > -1 && newPos > -1){
        if ( newPos > orgRightLength - 1 ) {
            return newPos - orgRightLength;
        }else{
            return newPos + revPos + 1;
        }
    }
    return newPos;
    
};

function findPos(nums, target, left, right) {
    if( left >= right ) return -1;
    let middlePos = left + (((right - left)/2)>>0);
    let middleVal = nums[middlePos];
    if( middleVal === target ) return middlePos;
    if( left === middlePos ) return -1;
    if (middleVal > target) {
        return findPos(nums, target, left, middlePos);
    }
    return findPos(nums, target, middlePos, right);
}

function findRevPos(nums, left, right){
    if( left >= right ) return -1;
    let middlePos = left + (((right - left)/2)>>0);
    let middleVal = nums[middlePos];
    if( middlePos < nums.length - 1 && middleVal > nums[middlePos+1] ) return middlePos;
    if( middlePos > 0 && middleVal < nums[middlePos-1] ) return middlePos-1;
    if( middleVal > nums[right - 1] ) return findRevPos(nums, middlePos, right);
    if( middleVal < nums[left] ) return findRevPos(nums, left, middlePos);
    return -1;
}

// @lc code=end

File: leetcode/33.搜索旋转排序数组.go
package leetcode

/*
 * @lc app=leetcode.cn id=33 lang=golang
 *
 * [33] 搜索旋转排序数组
 */

// @lc code=start
func search(nums []int, target int) int {
	left, right := 0, len(nums)-1
	for left <= right {
		mid := left + (right-left)/2
		if nums[mid] == target {
			return mid
		} else if goLeft(nums, target, left, right, mid) {
			right = mid - 1
		} else {
			left = mid + 1
		}
	}
	return -1
}

func goLeft(nums []int, target, left, right, mid int) bool {
	// 左边单调递增 且 target 小于 mid 大于 left
	con1 := nums[left] <= nums[mid] && target < nums[mid] && target >= nums[left]
	// 右边单调递增 且 target 小于 mid
	con2 := nums[mid] <= nums[right] && (target < nums[mid] || target > nums[right])
	return con1 || con2
}

// @lc code=end
